### PR TITLE
Support all SOVD resource collections in cyclic subscriptions

### LINKED ISSUE #253: cyclic-subscriptions/ should support all SOVD resource collections and vendor extensions
## Summary

Cyclic subscriptions are currently hard-coded to work only with `/data/` resources. Any other SOVD resource collection (faults, operations, configurations, communication-logs) or vendor extension (`x-medkit-*`) is rejected.

This means vendor resources registered by plugins via `register_capability()` can't be subscribed to, even though SOVD treats cyclic subscriptions as a generic mechanism for any resource on an entity path.

## What's hard-coded

Three places in `cyclic_subscription_handlers.cpp` restrict subscriptions to `/data/`:

1. **`parse_resource_uri()` regex** (line 442) - only matches `/api/v1/{apps|components}/{id}/data/{path}`:
   ```cpp
   static const std::regex resource_regex(R"(^/api/v1/(?:apps|components)/[^/]+/data(/.*))")
   ```

2. **Prefix validation** (line 118) - rejects any resource URI that doesn't start with `/data/`:
   ```cpp
   std::string expected_prefix = "/api/v1/" + entity_type + "/" + entity_id + "/data/";
   ```

3. **Entity type restriction** - routes are only registered for `apps` and `components`, not `functions`.

Additionally, `subscription_manager.hpp` uses `topic_name` for the extracted resource path, which assumes the resource is always a ROS 2 topic.

## Proposed solution (optional)

Make the subscription handler resource-agnostic:

- **`parse_resource_uri()`** should extract both the collection name and the resource path from the URI, not assume `/data/`
- **Prefix validation** should check that the resource URI matches the entity from the route, without assuming which collection it belongs to
- **Validation**: use `EntityCapabilities::supports_collection()` (for standard collections) or check registered plugin capabilities (for vendor extensions) to determine if the resource is subscribable
- **Rename** `topic_name` to `resource_path` in `CyclicSubscriptionInfo` for clarity
- **Route registration** should cover all entity types that support cyclic subscriptions per SOVD Table 8

The SSE streaming logic itself (`set_chunked_content_provider`, EventEnvelope format, replay buffer) is generic and shouldn't need changes - only the resource resolution and validation layer.

## Additional context (optional)

SOVD spec (ISO 17978-3, Section 7 - cyclic subscriptions) defines cyclic subscriptions as applicable to any resource on an entity, with `x-sovd-cyclic-subscription-supported: true` in the capability description indicating support. The current implementation doesn't check this flag or delegate to the resource handler for data retrieval - it goes directly to `NativeTopicSampler`, which only works for `/data/` (ROS 2 topics).

For non-data resources, the streaming callback would need to delegate to the appropriate handler/manager (e.g., FaultManager for faults, ConfigurationManager for configurations, plugin routes for vendor extensions) instead of sampling a ROS 2 topic directly.

### PR DESCRIPTION
**Pull request by bburda**
## Summary

Make cyclic subscriptions work with any SOVD resource collection (data, faults, configurations, communication-logs) and vendor extensions (`x-*` prefix), instead of being hard-coded to `/data/` only.

Key changes:
- **ResourceSamplerRegistry** - thread-safe registry mapping collection names to sampler functions. Built-in samplers (data, faults, configurations, communication-logs) use last-write-wins; plugins must use `x-` prefix
- **SubscriptionTransportProvider / TransportRegistry** - pluggable transport interface (SSE built-in, extensible to MQTT/WebSocket/Zenoh) with protocol-keyed registry
- **SseTransportProvider** - SSE streaming logic extracted from handler into a standalone transport provider with exception-safe sampler invocation
- **Generalized `parse_resource_uri()`** - extracts entity type, entity ID, collection, and resource path from any valid SOVD resource URI with path traversal protection
- **`CyclicSubscriptionInfo`** - replaced `topic_name` with `collection` + `resource_path` fields
- **`on_removed` callback** - SubscriptionManager lifecycle hook for transport cleanup on remove/expire/shutdown
- **Proper shutdown ordering** - REST stop -> transport stop -> sub_mgr shutdown -> plugin shutdown -> destruction
- **PluginManager** extensions - `register_resource_sampler()` and `register_transport()` for plugin-provided collections and transports

---

## Issue

- closes #253

---

## Type

- [ ] Bug fix
- [x] New feature or tests
- [ ] Breaking change
- [ ] Documentation only

---

## Testing

- 4 new unit test files (22 new test cases):
  - `test_resource_sampler_registry.cpp` - 7 tests (register, lookup, builtin overwrite, plugin reject, concurrent access)
  - `test_transport_registry.cpp` - 3 tests (register/lookup, missing transport, shutdown_all)
  - `test_cyclic_subscription_handlers.cpp` - 11 new ParseResourceUri tests (all collections, vendor extensions, entity types, path traversal, invalid URIs)
  - `test_subscription_manager.cpp` - 3 new on_removed callback tests (remove, expire, shutdown)
- 1 new integration test file:
  - `test_multi_collection_subscriptions.test.py` - 8 tests (data/faults/configurations create, unsupported collection, invalid URI, entity mismatch, unsupported protocol, path traversal)
- All 870 unit tests pass, all linters pass (clang-format, clang-tidy, ament)

---

## Checklist

- [ ] Breaking changes are clearly described (and announced in docs / changelog if needed)
- [x] Tests were added or updated if needed
- [ ] Docs were updated if behavior or public API changed

### CODE REVIEW

File: src/ros2_medkit_gateway/src/subscription_manager.cpp [hunk 1]
```diff
@@ -213,11 +227,21 @@ size_t SubscriptionManager::max_subscriptions() const {
...
+      it->second->cv.notify_all();
+      if (on_removed_) {
+        std::lock_guard<std::mutex> sub_lock(it->second->mtx);
+        all_infos.push_back(it->second->info);
+      }
+      it = subscriptions_.erase(it);
+    }
+  }
+  for (const auto & info : all_infos) {
+    on_removed_(info);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/subscription_manager.cpp]: Fixed in f781c62 - added `if (on_removed_)` guard around the loop in `shutdown()`, consistent with `remove()` and `cleanup_expired()`.

File: src/ros2_medkit_gateway/src/http/handlers/cyclic_subscription_handlers.cpp [hunk 2]
```diff
@@ -289,121 +352,31 @@ void CyclicSubscriptionHandlers::handle_events(const httplib::Request & req, htt
...
-      },
-      [this, captured_sub_id](bool /*success*/) {
-        client_tracker_->disconnect();
-        RCLCPP_DEBUG(rclcpp::get_logger("rest_server"), "SSE cyclic subscription stream disconnected: %s",
-                     captured_sub_id.c_str());
-      });
+  if (!transport->handle_client_connect(sub_id, req, res)) {
+    HandlerContext::send_error(res, 400, ERR_X_MEDKIT_UNSUPPORTED_PROTOCOL,
+                               "Use event_source URI for protocol '" + sub->protocol + "'",
+                               {{"protocol", sub->protocol}});
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/cyclic_subscription_handlers.cpp]: Fixed in f781c62 - changed to `ERR_RESOURCE_NOT_FOUND` with message "Subscription stream not found".

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/handlers/cyclic_subscription_handlers.hpp [hunk 3]
```diff
@@ -64,19 +76,22 @@ class CyclicSubscriptionHandlers {
...
-
   HandlerContext & ctx_;
   SubscriptionManager & sub_mgr_;
   std::shared_ptr<SSEClientTracker> client_tracker_;
+  ResourceSamplerRegistry & sampler_registry_;
+  TransportRegistry & transport_registry_;
+  int max_duration_sec_;
 
   /// Keepalive interval for SSE streams
   static constexpr int kKeepaliveIntervalSec = 15;
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/handlers/cyclic_subscription_handlers.hpp]: Fixed in f781c62 - removed dead `kKeepaliveIntervalSec` from `CyclicSubscriptionHandlers`. The constant lives in `SseTransportProvider` where it's actually used.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/handlers/cyclic_subscription_handlers.hpp [hunk 4]
```diff
@@ -41,7 +51,9 @@ namespace handlers {
 class CyclicSubscriptionHandlers {
  public:
   CyclicSubscriptionHandlers(HandlerContext & ctx, SubscriptionManager & sub_mgr,
-                             std::shared_ptr<SSEClientTracker> client_tracker);
+                             std::shared_ptr<SSEClientTracker> client_tracker,
+                             ResourceSamplerRegistry & sampler_registry, TransportRegistry & transport_registry,
+                             int max_duration_sec);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/handlers/cyclic_subscription_handlers.hpp:53]: Fixed in f781c62 - removed `client_tracker_` member, constructor parameter, and the `sse_client_tracker.hpp` include from `CyclicSubscriptionHandlers`. Also removed `<memory>` include that was only needed for the `shared_ptr`. Updated the call site in `rest_server.cpp`.

File: src/ros2_medkit_gateway/src/gateway_node.cpp [hunk 5]
```diff
@@ -556,10 +686,18 @@ GatewayNode::GatewayNode() : Node("ros2_medkit_gateway") {
...
+  // 2. Shutdown subscriptions via transport registry (calls sub_mgr.shutdown(),
+  //    which triggers on_removed -> transport->stop() for each active subscription)
+  if (transport_registry_) {
+    transport_registry_->shutdown_all(*subscription_mgr_);
+  }
+  // 3. Shutdown plugins
   if (plugin_mgr_) {
     plugin_mgr_->shutdown_all();
   }
+  // 5. Normal member destruction (managers safe - all transports stopped)
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/gateway_node.cpp]: Fixed in f781c62 - renumbered to step 4.

File: docs/tutorials/plugin-system.rst [hunk 6]
```diff
@@ -295,6 +295,41 @@ For entity-scoped endpoints, register a matching capability via ``register_capab
...
+
+.. code-block:: cpp
+
+   auto& transport_registry = ctx_->get_transport_registry();
+   transport_registry.register_transport(
+     std::make_unique<MqttTransportProvider>(mqtt_client_));
+
+The transport must implement ``SubscriptionTransportProvider`` (``start``, ``stop``,
+``notify_update``, ``protocol()``). Clients specify the protocol in the subscription
+creation request.
```
[the remark this answers is not in the document]
Comment (bburda) [reply docs/tutorials/plugin-system.rst]: Fixed in f781c62 - updated docs to show that sampler/transport registration uses `GatewayNode` API (`node->get_sampler_registry()`, `node->get_transport_registry()`), not `PluginContext`. The `PluginContext` interface does not expose these registries.

File: src/ros2_medkit_gateway/src/subscription_transport.cpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  transports_[proto] = std::move(provider);
+}
+
+SubscriptionTransportProvider * TransportRegistry::get_transport(const std::string & protocol) const {
+  auto it = transports_.find(protocol);
+  if (it == transports_.end()) {
+    return nullptr;
+  }
+  return it->second.get();
+}
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/subscription_transport.cpp:37]: Already addressed - `TransportRegistry` uses `std::shared_mutex` (see `subscription_transport.hpp:78`). `register_transport()` takes a `unique_lock` and `get_transport()`/`has_transport()` take `shared_lock`.

File: src/ros2_medkit_gateway/src/subscription_transport.cpp [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ros2_medkit_gateway/subscription_transport.hpp"
+
+namespace ros2_medkit_gateway {
+
+void TransportRegistry::register_transport(std::unique_ptr<SubscriptionTransportProvider> provider) {
+  auto proto = provider->protocol();
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/subscription_transport.cpp:23]: Already addressed - `register_transport()` throws `std::runtime_error` on duplicate protocol registration (see `subscription_transport.cpp:24-25`).

File: src/ros2_medkit_gateway/src/http/handlers/sse_transport_provider.cpp [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+void SseTransportProvider::notify_update(const std::string & sub_id) {
+  // SSE transport re-reads from sub_mgr_ each iteration, so no-op here.
+  (void)sub_id;
+}
+
+void SseTransportProvider::stop(const std::string & sub_id) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  streams_.erase(sub_id);
+  // The SSE loop detects inactive via sub_mgr_.is_active() and exits.
+}
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/sse_transport_provider.cpp:62]: The `stop()` method works correctly via the shutdown sequence: `on_removed` callback (set by `TransportRegistry`) is called after `state->active` is set to `false`. The SSE loop checks `sub_mgr_.is_active()` each iteration, which returns `false` once `active` is cleared. Erasing from `streams_` prevents new client connections. The comments in `stop()` explain this flow.

File: src/ros2_medkit_gateway/src/resource_sampler.cpp [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <stdexcept>
+
+namespace ros2_medkit_gateway {
+
+void ResourceSamplerRegistry::register_sampler(const std::string & collection, ResourceSamplerFn fn, bool is_builtin) {
+  std::unique_lock lock(mutex_);
+
+  if (!is_builtin) {
+    if (collection.substr(0, 2) != "x-") {
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/resource_sampler.cpp]: `collection` is extracted from a regex match group that requires at least one character (`([^/]+)`), so it cannot be empty when reaching the `substr` call. The regex guarantees `collection.size() >= 1`. For `size() == 1`, `substr(0, 2)` returns the single character (no throw - `substr` with `count > size()-pos` is valid per the standard, it just returns fewer chars). So this is safe as-is.

File: src/ros2_medkit_gateway/src/http/handlers/cyclic_subscription_handlers.cpp [hunk 11]
```diff
@@ -436,15 +387,27 @@ std::string CyclicSubscriptionHandlers::extract_entity_type(const httplib::Reque
...
+
+  ParsedResourceUri parsed;
+  parsed.entity_type = match[1].str();
+  parsed.entity_id = match[2].str();
+  parsed.collection = match[3].str();
+  parsed.resource_path = match[4].matched ? match[4].str() : "";
+
+  // Security: reject path traversal in resource_path
+  if (parsed.resource_path.find("/..") != std::string::npos || parsed.resource_path.find("../") != std::string::npos) {
+    return tl::make_unexpected("Resource path must not contain '..'");
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/cyclic_subscription_handlers.cpp]: The traversal check uses a regex that matches `..` as a full path segment: `(^|/)\\.\\.(/|$)`. This correctly rejects `/a/../b` and `/a/..` while allowing `/..foo` (tested in `BenignDoubleDotInSegmentAllowed`). The test suite covers all edge cases.

File: src/ros2_medkit_gateway/src/http/handlers/cyclic_subscription_handlers.cpp [hunk 12]
```diff
@@ -436,15 +387,27 @@ std::string CyclicSubscriptionHandlers::extract_entity_type(const httplib::Reque
...
+  }
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/cyclic_subscription_handlers.cpp]: Same as above - the regex `(^|/)\\.\\.(/|$)` checks segment boundaries correctly. `/...` would NOT be rejected because `...` \!= `..` as a segment. Test `BenignDoubleDotInSegmentAllowed` verifies `/..foo` is allowed.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/subscription_manager.hpp [hunk 13]
```diff
@@ -112,6 +115,9 @@ class SubscriptionManager {
   /// Signal shutdown — wakes all waiting streams
   void shutdown();
 
+  /// Set callback invoked when a subscription is removed (remove, cleanup_expired, shutdown)
+  void set_on_removed(std::function<void(const CyclicSubscriptionInfo &)> callback);
+
   /**
    * @brief Wait for an update or removal on the given subscription.
    *
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/subscription_manager.hpp]: `set_on_removed()` is called once during initialization (in `GatewayNode` constructor, before any HTTP server threads start). The `assert(\!has_subscriptions)` enforces this contract. Since `on_removed_` is set-once-before-threads, there is no data race. Adding mutex protection for a set-once callback would add unnecessary overhead to every `remove()`/`cleanup_expired()` call.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/subscription_manager.hpp [hunk 14]
```diff
@@ -134,6 +140,7 @@ class SubscriptionManager {
   size_t max_subscriptions_;
   std::atomic<uint64_t> next_id_{1};
   std::atomic<bool> shutdown_flag_{false};
+  std::function<void(const CyclicSubscriptionInfo &)> on_removed_;
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/subscription_manager.hpp:144]: Same as above - `set_on_removed()` is set-once during initialization before threads start. The assert enforces this contract.

File: src/ros2_medkit_gateway/src/http/handlers/cyclic_subscription_handlers.cpp [hunk 15]
```diff
@@ -103,35 +108,70 @@ void CyclicSubscriptionHandlers::handle_create(const httplib::Request & req, htt
...
+    HandlerContext::send_error(res, 400, ERR_X_MEDKIT_COLLECTION_NOT_AVAILABLE,
+                               "No data provider for collection '" + parsed->collection + "'",
+                               {{"collection", parsed->collection}});
+    return;
+  }
+
   // Create subscription
-  auto result = sub_mgr_.create(entity_id, entity_type, resource, *topic_result, protocol, interval, duration);
+  auto result = sub_mgr_.create(entity_id, entity_type, resource, parsed->collection, parsed->resource_path, protocol,
+                                interval, duration);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/cyclic_subscription_handlers.cpp:179]: Good observation. The `data` collection sampler itself returns an error when `resource_path` is empty (no topic specified), so the subscription would stream error frames. This is valid behavior - early validation in `parse_resource_uri` would be an optimization but not a correctness issue. Will consider adding this validation as a follow-up enhancement.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/subscription_transport.hpp [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+/// Registry of transport providers keyed by protocol name.
+class TransportRegistry {
+ public:
+  void register_transport(std::unique_ptr<SubscriptionTransportProvider> provider);
+  SubscriptionTransportProvider * get_transport(const std::string & protocol) const;
+  bool has_transport(const std::string & protocol) const;
+
+  /// Stop all active subscriptions across all transports.
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/subscription_transport.hpp]: The docstring already explains this: "Shutdown all subscriptions by delegating to SubscriptionManager::shutdown(). Transport cleanup happens via the manager's on_removed callback, which calls SubscriptionTransportProvider::stop() for each active subscription." Updated the docstring in f781c62 to be even clearer.

File: src/ros2_medkit_gateway/src/gateway_node.cpp [hunk 17]
```diff
@@ -489,10 +618,21 @@ GatewayNode::GatewayNode() : Node("ros2_medkit_gateway") {
...
+  }
+  // 3. Shutdown subscription manager (marks all inactive, triggers on_removed for stragglers)
+  if (subscription_mgr_) {
+    subscription_mgr_->shutdown();
+  }
+  // 4. Shutdown plugins
   if (plugin_mgr_) {
     plugin_mgr_->shutdown_all();
   }
+  // 5. Normal member destruction (managers safe - all transports stopped)
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/gateway_node.cpp]: `shutdown_all()` calls `sub_mgr.shutdown()` which triggers `on_removed` -> `transport->stop()` for each subscription and clears the map. The destructor's `subscription_mgr_->shutdown()` is a no-op on second call (map is already empty, `shutdown_flag_` already set). This is intentional defensive coding - the destructor ensures cleanup even if `shutdown_all` was never called. No double-trigger of callbacks since there are no subscriptions left.

File: src/ros2_medkit_gateway/src/gateway_node.cpp [hunk 18]
```diff
@@ -517,6 +529,144 @@ GatewayNode::GatewayNode() : Node("ros2_medkit_gateway") {
...
+  // Register built-in resource samplers
+  sampler_registry_->register_sampler(
+      "data",
+      [this](const std::string & /*entity_id*/,
+             const std::string & resource_path) -> tl::expected<nlohmann::json, std::string> {
+        auto * dam = get_data_access_manager();
+        if (!dam) {
+          return tl::make_unexpected(std::string("DataAccessManager not available"));
+        }
+        auto * native_sampler = dam->get_native_sampler();
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/gateway_node.cpp:541]: **Major: Null pointer dereference risk**

`dam->get_native_sampler()` result is used directly on the next line without a null check. If `get_native_sampler()` returns nullptr, this crashes.

```cpp
auto * native_sampler = dam->get_native_sampler();
if (!native_sampler) {
  return tl::make_unexpected(std::string("native sampler unavailable"));
}
auto sample = native_sampler->sample_topic(resource_path, dam->get_topic_sample_timeout());
```

File: src/ros2_medkit_gateway/src/http/handlers/sse_transport_provider.cpp [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+  return std::string(API_BASE_PATH) + "/" + info.entity_type + "/" + info.entity_id + "/cyclic-subscriptions/" +
+         info.id + "/events";
+}
+
+void SseTransportProvider::notify_update(const std::string & sub_id) {
+  // SSE transport re-reads from sub_mgr_ each iteration, so no-op here.
+  (void)sub_id;
+}
+
+void SseTransportProvider::stop(const std::string & sub_id) {
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/http/handlers/sse_transport_provider.cpp:55]: **Major: `stop()` violates the synchronous contract**

The `SubscriptionTransportProvider` interface says:
> MUST be synchronous - return only when all threads stopped.

But this implementation just erases from `streams_` and returns. The in-flight SSE streaming loop in `handle_client_connect` continues running (holding a copy of the sampler lambda that captures `GatewayNode*`) until the next `is_active()` check.

This works today because `stop_rest_server()` joins HTTP threads before `shutdown_all()` is called — but the contract violation is real, and a future non-HTTP transport could hit use-after-free.

**Suggestion:** Either:
1. Add a synchronization barrier (e.g. `std::condition_variable` + counter of active loops, wait until zero), or
2. Weaken the interface contract to "best-effort" and document the shutdown ordering dependency explicitly in both the interface comment and `GatewayNode::~GatewayNode()`.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/models/entity_types.hpp [hunk 20]
```diff
@@ -49,7 +49,8 @@ enum class ResourceCollection {
   LOCKS,                 ///< Lock resources (lifecycle states)
   MODES,                 ///< Mode resources (node modes)
   CYCLIC_SUBSCRIPTIONS,  ///< Cyclic subscriptions (topic polling)
-  COMMUNICATION_LOGS,    ///< Communication logs (/rosout filtered)
+  LOGS,                  ///< Application log entries (/rosout)
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/include/ros2_medkit_gateway/models/entity_types.hpp:52]: **Major: Breaking REST API change - undocumented**

Adding `LOGS` and repurposing `COMMUNICATION_LOGS` changes the capabilities array returned by the REST API. Clients (sovd_web_ui, ros2_medkit_mcp) that check for `"communication-logs"` to mean application logs will break — the API now returns `"logs"` instead.

The PR checklist has "Breaking changes are clearly described" unchecked.

**Suggestion:** Either:
1. Document this as a breaking change + update downstream clients in the same release, or
2. Keep `"communication-logs"` in capability sets alongside `"logs"` for one release cycle as a deprecation path.

File: src/ros2_medkit_gateway/src/gateway_node.cpp [hunk 21]
```diff
@@ -89,6 +91,7 @@ GatewayNode::GatewayNode() : Node("ros2_medkit_gateway") {
   // SSE (Server-Sent Events) parameters
   declare_parameter("sse.max_clients", 10);         // Limit concurrent SSE connections to prevent resource exhaustion
   declare_parameter("sse.max_subscriptions", 100);  // Maximum active cyclic subscriptions across all entities
+  declare_parameter("sse.max_duration_sec", 3600);  // Maximum subscription duration in seconds (1 hour default)
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/gateway_node.cpp:94]: **Minor: No bounds validation on `max_duration_sec`**

This parameter is read via `as_int()` with no minimum check. A negative or zero value would make every subscription create fail with HTTP 400 (since `duration > max_duration_sec_` is always true for any valid `duration > 0`).

```cpp
auto max_dur = get_parameter("sse.max_duration_sec").as_int();
if (max_dur <= 0) {
  RCLCPP_WARN(get_logger(), "sse.max_duration_sec must be > 0, using default 3600");
  max_dur = 3600;
}
```

File: src/ros2_medkit_gateway/src/gateway_node.cpp [hunk 22]
```diff
@@ -517,6 +529,144 @@ GatewayNode::GatewayNode() : Node("ros2_medkit_gateway") {
...
+      [this](const std::string & entity_id,
+             const std::string & /*resource_path*/) -> tl::expected<nlohmann::json, std::string> {
+        auto * fault_mgr = get_fault_manager();
+        if (!fault_mgr) {
+          return tl::make_unexpected(std::string("FaultManager not available"));
+        }
+        // Determine source_id for fault filtering based on entity type
+        const auto & cache = get_thread_safe_cache();
+        auto entity_ref = cache.find_entity(entity_id);
+        std::string source_id;
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/gateway_node.cpp:567]: **Minor: Race between `find_entity` and `get_app`/`get_component`**

If `entity_ref` is found but `cache.get_app(entity_id)` returns `nullopt` (entity removed between the two calls), `source_id` stays empty → `list_faults("")` returns **all** faults across all entities instead of scoping to the requested entity.

Not a crash, but a correctness issue — the client gets a superset of faults. Consider returning an error if the entity vanished:

```cpp
if (entity_ref && source_id.empty() && entity_ref->type != SovdEntityType::AREA) {
  return tl::make_unexpected(std::string("Entity no longer available: " + entity_id));
}
```

File: src/ros2_medkit_gateway/src/gateway_node.cpp (hunk 18 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/gateway_node.cpp:541]: Fixed in 6914c9d - added null guard for `native_sampler`. In practice `native_sampler_` is always initialized in `DataAccessManager` constructor (`make_unique`), but the defensive check is cheap and consistent with how `dam` itself is checked.

File: src/ros2_medkit_gateway/src/http/handlers/sse_transport_provider.cpp (hunk 19 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/sse_transport_provider.cpp:55]: Fixed in 6914c9d - went with option 2: updated the `SubscriptionTransportProvider::stop()` docstring to document the shutdown ordering dependency. For SSE, `stop_rest_server()` joins HTTP threads before `shutdown_all()` triggers `stop()`, so the loop exits naturally. Non-HTTP transports must ensure synchronous stop themselves.

File: src/ros2_medkit_gateway/src/gateway_node.cpp (hunk 21 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/gateway_node.cpp:94]: Fixed in 6914c9d - added bounds check in `rest_server.cpp`. If `sse.max_duration_sec <= 0`, logs a warning and defaults to 3600.

File: src/ros2_medkit_gateway/src/gateway_node.cpp (hunk 22 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/gateway_node.cpp:567]: Fixed in 6914c9d - added TOCTOU guard: if `source_id` is empty for a non-AREA/FUNCTION entity, returns an error instead of falling through to `list_faults("")` (which would return all faults).

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/models/entity_types.hpp (hunk 20 above)
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/models/entity_types.hpp:52]: This is a bug I've introduced in the #245 it should be log instead of comm logs
